hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/upgrade/BasicUpgradeFinalizer.java

```diff
@@ -57,18 +57,39 @@
   protected V versionManager;
   protected String clientID;
   protected T component;
+  protected UpgradeFinalizationExecutor finalizationExecutor;
 
   private Queue<String> msgs = new ConcurrentLinkedQueue<>();
   protected boolean isDone = false;
 
   public BasicUpgradeFinalizer(V versionManager) {
     this.versionManager = versionManager;
+    this.finalizationExecutor =
+        new UpgradeFinalizationExecutor();
+  }
+
+  @Override
+  public void setFinalizationExecutor(UpgradeFinalizationExecutor executor) {
+    finalizationExecutor = executor;
+  }
+
+  @Override
+  public UpgradeFinalizationExecutor getFinalizationExecutor() {
+    return finalizationExecutor;
   }
 
   public boolean isFinalizationDone() {
     return isDone;
   }
 
+  public void markFinalizationDone() {
+    isDone = true;
```

Comment:
Can we not have **state** here? Why cannot we refer to the version manager to figure out that finalization is done or not?

The way I see it, the version manager maintains state. The upgrade finalizer provides component specific upgrade finalization implementations (without state), and the executor uses both of them to orchestrate the upgrade.

Reply:
Currently OMUpgrade is the path that makes use of it and for some OM upgrade unit tests. If we refactor OM upgrade, we can get rid of this.

Reply:
Created HDDS-5108 to track this work item.